README.md
# Cavendish-MW

This is the README file for Cavendish-MW 0.3, a skin for MediaWiki. This skin
is based on Mozilla's Cavendish skin for MediaWiki.

Cavendish-MW 0.3 has been tested with MediaWiki 1.21.0. It might not work
well with other versions of MediaWiki.

## Installation Instructions

To install the Cavendish-MW skin, extract the files from the archive to the
"skins" folder of your MediaWiki installation.

Alternatively, you can use Git to clone the entire repository (the advantage
is that you can easily update the skin by running the command `git pull`):

    cd skins/
    git clone https://github.com/figure002/cavendishmw.git

You should now have the directory `skins/cavendishmw/` in your MediaWiki
installation.

Now open LocalSettings.php and add the following line somewhere at the end of
the file:

    require_once("$IP/skins/cavendishmw/cavendishmw.php");

That's it.

On your MediaWiki Preferences page, click on the Appearance tab. Under "Skin"
there should be an option "Cavendish-MW". Select the skin you would like to use
for your account and click Save.

## Configuration

MediaWiki and the Cavendish-MW skin can be configured by editing the file
LocalSettings.php in the root folder of your MediaWiki installation. Open this
file with a text editor.

### Set as default

If you want to set the default skin for all users to Cavendish-MW, set the
variable `$wgDefaultSkin` to 'cavendishmw'. The value must be in lowercase,

    $wgDefaultSkin = 'cavendishmw';

### Change the header logo

To change the header logo, change the variable `$wgLogo`. Set it to the path
of the logo file.

### Cavendish-MW variables

You can use the following variables to further customize Cavendish-MW. The
values shown below are the default values.

    // Set to 'false' to hide the site name in the header.
    $cavendishShowSitename = true;

    // The indentation for the site name. You may alter the value depending on
    // the size of the logo you use.
    $cavendishSitenameIndent = '2em';

    // The width of the site name area. Set this to a fixed value if your logo
    // exeeds the width of the area.
    $cavendishLogoWidth = 'auto';

    // The height of the site name area. Set this to a fixed value if your logo
    // exeeds the height of the area.
    $cavendishLogoHeight = 'auto';

## License

This work is licensed under the [Creative Commons Attribution-ShareAlike 3.0 Unported License](http://creativecommons.org/licenses/by-sa/3.0/).


HISTORY.md
Change notes from older releases.

## Cavendish-MW 0.3.2

* (merge 3) Small tweak for English as well as adding German translation.

  The English language tweak is to adhere the MediaWiki jargon as well as to
  i18n habits (no fullstop in description on "Special:Version"). Credits to
  Karsten Hoffmeyer.

## Cavendish-MW 0.3.1

* Fixed the display of the non-breaking space in the header link when the site
  name is hidden.

## Cavendish-MW 0.3.0

Updated for MediaWiki 1.21.0.

* Rewrite of the skin.

  The following tutorial was used:
  <https://www.mediawiki.org/wiki/Manual:Skinning/Tutorial>

  The `CavendishMWTemplate` class of CavendishMW.skin.php is based on the
  `MonoBookTemplate` class of the Monobook skin. The original stylesheets for
  Cavendish-MW were reused and updated.

  The Cavendish-MW skin now works as an extension and one needs to put
  `require_once("$IP/skins/cavendishmw/cavendishmw.php");` in their
  LocalSettings.php file.

* (bug 2) The ampersand symbol is now displayed correctly in page titles.


cavendishmw.php
<?php
/**
 * Cavendish-MW: Branch of the Mozilla Cavendish MediaWiki skin with some
 * improvements.
 *
 * @file
 * @ingroup Skins
 * @author Serrano Pereira (serrano.pereira@gmail.com)
 * @license http://creativecommons.org/licenses/by-sa/3.0/ Creative Commons Attribution-ShareAlike 3.0 Unported License
 */

if( !defined( 'MEDIAWIKI' ) ) die( "This is an extension to the MediaWiki package and cannot be run standalone." );

$wgExtensionCredits['skin'][] = array(
    'path' => __FILE__,
    'name' => 'Cavendish-MW',
    'version' => '0.3.2',
    'url' => "http://sourceforge.net/projects/cavendishmw/",
    'author' => array('Serrano Pereira'),
    'descriptionmsg' => 'cavendishmw-desc',
);

$wgValidSkinNames['cavendishmw'] = 'CavendishMW';
$wgAutoloadClasses['SkinCavendishMW'] = dirname(__FILE__).'/CavendishMW.skin.php';
$wgExtensionMessagesFiles['CavendishMW'] = dirname(__FILE__).'/CavendishMW.i18n.php';

$wgResourceModules['skins.cavendishmw'] = array(
    'styles' => array(
        'cavendishmw/styles/monobook.css' => array( 'media' => 'screen' ),
        'cavendishmw/styles/template.css' => array( 'media' => 'screen' ),
        'cavendishmw/styles/basetemplate.css' => array( 'media' => 'screen' ),
        'cavendishmw/styles/cavendish.css' => array( 'media' => 'screen' ),
        'cavendishmw/styles/fixes.css' => array( 'media' => 'screen' ),
    ),
    'remoteBasePath' => &$GLOBALS['wgStylePath'],
    'localBasePath' => &$GLOBALS['wgStyleDirectory'],
);


CavendishMW.skin.php
<?php
/**
 * Cavendish-MW: Branch of the Mozilla Cavendish MediaWiki skin with some
 * improvements.
 *
 * Loosely based on the Cavendish style by Gabriel Wicke.
 *
 * This work is licensed under the Creative Commons Attribution-ShareAlike 3.0
 * Unported License. To view a copy of this license, visit
 * http://creativecommons.org/licenses/by-sa/3.0/ or send a letter to Creative
 * Commons, 171 Second Street, Suite 300, San Francisco, California, 94105, USA.
 *
 * @file
 * @ingroup Skins
 */

if( !defined( 'MEDIAWIKI' ) )
    die( -1 );

/**
 * Inherit main code from SkinTemplate, set the CSS and template filter.
 * @ingroup Skins
 */
class SkinCavendishMW extends SkinTemplate {
    var $skinname = 'cavendishmw', $stylename = 'cavendishmw',
        $template = 'CavendishMWTemplate', $useHeadElement = true;

    /**
     * Load skin and user CSS files in the correct order
     * @param $out OutputPage object
     */
    function setupSkinUserCss( OutputPage $out ) {
        parent::setupSkinUserCss( $out );
        // Use the monobook styles as the basis.
        //$out->addModuleStyles( 'skins.monobook' );
        // Load the cavendish stylesheets.
        $out->addModuleStyles( 'skins.cavendishmw' );
    }
}

/**
 * BaseTemplate class for CavendishMW skin
 * @ingroup Skins
 */
class CavendishMWTemplate extends BaseTemplate {

    /**
     * Outputs the entire contents of the page
     */
    public function execute() {
        global $cavendishShowSitename, $cavendishSitenameIndent,
            $cavendishLogoWidth, $cavendishLogoHeight;

        // Suppress warnings to prevent notices about missing indexes in $this->data
        wfSuppressWarnings();

        // Set CavendishMW specific variables.
        $cavendishShowSitename = isset($cavendishShowSitename) ? $cavendishShowSitename : true;
        $cavendishSitenameIndent = isset($cavendishSitenameIndent) ? $cavendishSitenameIndent : '2em';
        $cavendishLogoWidth = isset($cavendishLogoWidth) ? $cavendishLogoWidth : 'auto';
        $cavendishLogoHeight = isset($cavendishLogoHeight) ? $cavendishLogoHeight : 'auto';
        $this->set('sitenameindent', $cavendishSitenameIndent);
        $this->set('logowidth', $cavendishLogoWidth);
        $this->set('logoheight', $cavendishLogoHeight);

        // Print the HTML head
        $this->html( 'headelement' );
?>

<div id="container"> <!-- cavendishmw: s/globalWrapper/container/ -->

    <!-- <div id="mozilla-org"><a href="#">Mozilla Skin</a></div> -->
    <div id="header" class="noprint">
        <a name="top" id="contentTop"></a>

        <!-- Logo + site name -->
        <h1>
            <?php
            $contents = $cavendishShowSitename ? $this->data['sitename'] : "#nbsp";
            $link = Html::element( 'a', array(
                'href' => $this->data['nav_urls']['mainpage']['href'],
                'style' => "text-indent: {$this->data['sitenameindent']}; width: {$this->data['logowidth']}; height: {$this->data['logoheight']}; background: transparent url({$this->data['logopath']}) no-repeat scroll 5px -5px;"
            ) + Linker::tooltipAndAccesskeyAttribs('p-logo'), $contents );
            // Can't use &nbsp; in Html::element directly.
            echo str_replace("#nbsp", "&nbsp;", $link);
            ?>
        </h1>

        <!-- Content action buttons -->
        <?php $this->cactions(); ?>

        <!-- Search box -->
        <?php $this->searchBox(); ?>
    </div> <!-- End header div -->

    <div id="mBody">
        <div id="side" class="noprint" <?php $this->html('userlangattributes') ?>> <!-- cavendishmw: s/column-one/side/ -->
            <ul id="nav">
                <?php
                // Display Personal tools.
                $this->personalTools();
                // Display other Navigation blocks.
                $this->renderPortals( $this->data['sidebar'] );
                ?>
            </ul> <!-- /nav -->
        </div> <!-- /side -->

        <div id="mainContent"> <!-- cavendishmw: s/column-content/mainContent/ -->
            <?php if($this->data['sitenotice']) { ?><div id="siteNotice"><?php $this->html('sitenotice') ?></div><?php } ?>

            <h1 id="firstHeading" class="firstHeading" lang="<?php
                $this->data['pageLanguage'] = $this->getSkin()->getTitle()->getPageViewLanguage()->getCode();
                $this->html( 'pageLanguage' );
            ?>"><span dir="auto"><?php $this->html('title') ?></span></h1>
            <div id="bodyContent" class="mw-body">
                <div id="siteSub"><?php $this->msg('tagline') ?></div>
                <div id="contentSub"<?php $this->html('userlangattributes') ?>><?php $this->html('subtitle') ?></div>
            <?php if($this->data['undelete']) { ?>
                <div id="contentSub2"><?php $this->html('undelete') ?></div>
            <?php } ?><?php if($this->data['newtalk'] ) { ?>
                <div class="usermessage"><?php $this->html('newtalk')  ?></div>
            <?php } ?><?php if($this->data['showjumplinks']) { ?>
                <div id="jump-to-nav" class="mw-jump"><?php $this->msg('jumpto') ?> <a href="#column-one"><?php $this->msg('jumptonavigation') ?></a><?php $this->msg( 'comma-separator' ) ?><a href="#searchInput"><?php $this->msg('jumptosearch') ?></a></div>
            <?php } ?>

                <!-- start content -->
                <?php $this->html('bodytext') ?>
                <?php if($this->data['catlinks']) { $this->html('catlinks'); } ?>
                <!-- end content -->

                <?php if($this->data['dataAfterContent']) { $this->html ('dataAfterContent'); } ?>
                <div class="visualClear"></div>
            </div>
        </div> <!-- /mainContent -->
    </div> <!-- /mBody -->

    <div class="visualClear"></div>
    <?php
        $validFooterIcons = $this->getFooterIcons( "icononly" );
        $validFooterLinks = $this->getFooterLinks( "flat" ); // Additional footer links

        if ( count( $validFooterIcons ) + count( $validFooterLinks ) > 0 ) { ?>
    <div id="footer" role="contentinfo"<?php $this->html('userlangattributes') ?>>
    <?php
            $footerEnd = '</div>';
        } else {
            $footerEnd = '';
        }
        foreach ( $validFooterIcons as $blockName => $footerIcons ) { ?>
        <div id="f-<?php echo htmlspecialchars($blockName); ?>ico">
    <?php foreach ( $footerIcons as $icon ) { ?>
            <?php echo $this->getSkin()->makeFooterIcon( $icon ); ?>

    <?php }
    ?>
        </div>
    <?php }

            if ( count( $validFooterLinks ) > 0 ) {
    ?>    <ul id="f-list">
    <?php
                foreach( $validFooterLinks as $aLink ) { ?>
            <li id="<?php echo $aLink ?>"><?php $this->html($aLink) ?></li>
    <?php
                }
    ?>
        </ul>
    <?php    }
    echo $footerEnd;
    ?>

    </div>
<?php
        $this->printTrail();
        echo Html::closeElement( 'body' );
        echo Html::closeElement( 'html' );
        wfRestoreWarnings();
    } // end of execute() method

    /*************************************************************************************************/

    /**
     * @param $sidebar array
     */
    protected function renderPortals( $sidebar ) {
        if ( !isset( $sidebar['SEARCH'] ) ) $sidebar['SEARCH'] = true;
        if ( !isset( $sidebar['TOOLBOX'] ) ) $sidebar['TOOLBOX'] = true;
        if ( !isset( $sidebar['LANGUAGES'] ) ) $sidebar['LANGUAGES'] = true;

        foreach( $sidebar as $boxName => $content ) {
            if ( $content === false )
                continue;

            if ( $boxName == 'SEARCH' ) {
                // The searchbox is disabled, because we already have one in the header.
                // Uncomment the line below to enable it again.
                //$this->searchBox();
            } elseif ( $boxName == 'TOOLBOX' ) {
                $this->toolbox();
            } elseif ( $boxName == 'LANGUAGES' ) {
                $this->languageBox();
            } else {
                $this->customBox( $boxName, $content );
            }
        }
    }

    /**
     * Prints the search box.
     */
    function searchBox() {
        global $wgUseTwoButtonsSearchForm;
?>
    <form action="<?php $this->text('wgScript') ?>" id="searchform">
        <label for="searchInput"><?php $this->msg('search') ?></label>
        <input type='hidden' name="title" value="<?php $this->text('searchtitle') ?>"/>
        <?php echo $this->makeSearchInput(array( "id" => "searchInput" )); ?>

        <?php echo $this->makeSearchButton("go", array( "id" => "searchGoButton", "class" => "searchButton" ));
        if ($wgUseTwoButtonsSearchForm): ?>&#160;
        <?php echo $this->makeSearchButton("fulltext", array( "id" => "mw-searchButton", "class" => "searchButton" ));
        else: ?>

        <div><a href="<?php $this->text('searchaction') ?>" rel="search"><?php $this->msg('powersearch-legend') ?></a></div><?php
        endif; ?>
    </form>
<?php
    }

    /**
     * Prints the cactions bar.
     */
    function cactions() {
?>
        <ul><?php
            foreach($this->data['content_actions'] as $key => $tab) {
                echo '
            ' . $this->makeListItem( $key, $tab );
            } ?>

        </ul>
<?php
    }

    /**
     * Prints the personal tools.
     */
    function personalTools() {
?>
        <li><span><?php $this->msg('personaltools') ?></span>
            <ul<?php $this->html('userlangattributes') ?>>
            <?php foreach($this->getPersonalTools() as $key => $item) { ?>
                <?php echo $this->makeListItem($key, $item); ?>
            <?php } ?>
            </ul>
        </li>
<?php
    }
    /*************************************************************************************************/
    function toolbox() {
?>
    <li><span><?php $this->msg('toolbox') ?></span>
        <ul>
<?php
        foreach ( $this->getToolbox() as $key => $tbitem ) { ?>
                <?php echo $this->makeListItem($key, $tbitem); ?>

<?php
        }
        wfRunHooks( 'MonoBookTemplateToolboxEnd', array( &$this ) );
        wfRunHooks( 'SkinTemplateToolboxEnd', array( &$this, true ) );
?>
        </ul>
    </li>
<?php
    }

    /*************************************************************************************************/
    function languageBox() {
        if( $this->data['language_urls'] ) {
?>
    <li><span><?php $this->msg('otherlanguages') ?></span>
        <ul>
<?php        foreach($this->data['language_urls'] as $key => $langlink) { ?>
                <?php echo $this->makeListItem($key, $langlink); ?>

<?php        } ?>
        </ul>
    </li>
<?php
        }
    }

    /*************************************************************************************************/
    /**
     * @param $bar string
     * @param $cont array|string
     */
    function customBox( $bar, $cont ) {
?>
        <li><span><?php $msg = wfMessage( $bar ); echo htmlspecialchars( $msg->exists() ? $msg->text() : $bar ); ?></span>
<?php   if ( is_array( $cont ) ) { ?>
            <ul>
<?php             foreach($cont as $key => $val) { ?>
                <?php echo $this->makeListItem($key, $val); ?>

<?php            } ?>
            </ul>
<?php   } else {
            # allow raw HTML block to be defined by extensions
            print $cont;
        }
?>
        </li>
<?php
    }
} // end of class


CavendishMW.php
<?php
/**
 * Cavendish-MW - Branch of the Mozilla Cavendish MediaWiki skin with some
 * improvements.
 *
 * Loosely based on the Cavendish style by Gabriel Wicke.
 *
 * Modified 2012/05/08, Serrano Pereira <serrano.pereira@gmail.com>
 *
 * This work is licensed under the Creative Commons Attribution-ShareAlike 3.0
 * Unported License. To view a copy of this license, visit
 * http://creativecommons.org/licenses/by-sa/3.0/ or send a letter to Creative
 * Commons, 171 Second Street, Suite 300, San Francisco, California, 94105, USA.
 *
 */

if( !defined( 'MEDIAWIKI' ) )
	die( -1 );

/**
 * Inherit main code from SkinTemplate, set the CSS and template filter.
 * @todo document
 * @ingroup Skins
 */
class SkinCavendishMW extends SkinTemplate {
	var $skinname = 'cavendishmw', $stylename = 'cavendishmw',
		$template = 'CavendishMWTemplate', $useHeadElement = true;

	function setupSkinUserCss( OutputPage $out ) {
		global $wgHandheldStyle;

		parent::setupSkinUserCss( $out );

		// Append to the default screen common & print styles...
		$out->addStyle( 'cavendishmw/main.css', 'screen' );
	}

	// This line fixes a later bug in which $skin->tooltipAndAccesskey no longer
	// exist and is now Xml::expandAttributes(Linker::tooltipAndAccesskeyAttribs($value)).
	function tooltipAndAccesskey($value) {
        return Xml::expandAttributes(Linker::tooltipAndAccesskeyAttribs($value));
    }
}

/**
 * @todo document
 * @ingroup Skins
 */
class CavendishMWTemplate extends BaseTemplate {
	var $skin;

	/**
	 * Template filter callback for Cavendish skin.
	 * Takes an associative array of data set from a SkinTemplate-based
	 * class, and a wrapper for MediaWiki's localization database, and
	 * outputs a formatted page.
	 *
	 * @access private
	 */
	function execute() {
		global $wgRequest, $wgSitename,
            $cavendishShowSitename, $cavendishSitenameIndent,
            $cavendishLogoWidth, $cavendishLogoHeight;
		$this->skin = $skin = $this->data['skin'];
		$action = $wgRequest->getText( 'action' );

        // Set cavendishmw specific variables.
        $cavendishShowSitename = isset($cavendishShowSitename) ? $cavendishShowSitename : true;
        $cavendishSitenameIndent = isset($cavendishSitenameIndent) ? $cavendishSitenameIndent : '2em';
        $cavendishLogoWidth = isset($cavendishLogoWidth) ? $cavendishLogoWidth : 'auto';
        $cavendishLogoHeight = isset($cavendishLogoHeight) ? $cavendishLogoHeight : 'auto';
        $this->set('sitenameindent', $cavendishSitenameIndent);
        $this->set('logowidth', $cavendishLogoWidth);
        $this->set('logoheight', $cavendishLogoHeight);

        // Retrieve site name
        $this->set('sitename', $wgSitename);

		// Suppress warnings to prevent notices about missing indexes in $this->data
		wfSuppressWarnings();

        $this->html( 'headelement' );

?><div id="internal"></div> <!-- cavendish-mw -->
<div id="container"> <!-- cavendish-mw / default: globalWrapper -->

    <!-- <div id="mozilla-org"><a href="https://sourceforge.net/projects/cavendishmw/">Mozilla Skin</a></div> -->
    <?php if($this->data['sitenotice']) { ?><div id="siteNotice"><?php $this->html('sitenotice') ?></div><?php } ?>

	<div id="header" class="noprint">
		<a name="top" id="contentTop"></a>
		<h1><a style="text-indent: <?php $this->text('sitenameindent'); ?>; width: <?php $this->text('logowidth'); ?>; height: <?php $this->text('logoheight'); ?>; background: transparent url(<?php $this->text('logopath') ?>) no-repeat scroll 5px -5px;);" href="<?php echo htmlspecialchars($this->data['nav_urls']['mainpage']['href'])?>"<?php echo $skin->tooltipAndAccesskey('p-logo') ?>><?php if ($cavendishShowSitename) { $this->text('sitename'); } else { print "&nbsp;"; } ?></a></h1>

		<ul> <!-- Start of content action buttons -->
            <?php foreach($this->data['content_actions'] as $key => $tab) {
		        echo '
	         <li id="' . Sanitizer::escapeId( "ca-$key" ) . '"';
		        if( $tab['class'] ) {
			        echo ' class="'.htmlspecialchars($tab['class']).'"';
		        }
		        echo'><a href="'.htmlspecialchars($tab['href']).'"';
		        # We don't want to give the watch tab an accesskey if the
		        # page is being edited, because that conflicts with the
		        # accesskey on the watch checkbox.  We also don't want to
		        # give the edit tab an accesskey, because that's fairly su-
		        # perfluous and conflicts with an accesskey (Ctrl-E) often
		        # used for editing in Safari.
	         	if( in_array( $action, array( 'edit', 'submit' ) )
	         	&& in_array( $key, array( 'edit', 'watch', 'unwatch' ))) {
	         		echo $skin->tooltip( "ca-$key" );
	         	} else {
	         		echo $skin->tooltipAndAccesskey( "ca-$key" );
	         	}
	         	echo '>'.htmlspecialchars($tab['text']).'</a></li>';
            } ?>
		</ul>

		<form action="<?php $this->text('searchaction') ?>" id="searchform">
			<div>
			<label for="searchInput"><?php $this->msg('search') ?></label>
			<input id="searchInput" name="search" type="text" <?php echo $this->skin->tooltipAndAccesskey('search'); if( isset( $this->data['search'] ) ) { ?> value="<?php $this->text('search') ?>"<?php } ?> />
			<input type='submit' name="go" class="searchButton" id="searchGoButton"	value="<?php $this->msg('searcharticle') ?>"<?php echo $this->skin->tooltipAndAccesskey( 'search-go' ); ?> />
			<input type='submit' name="fulltext" class="searchButton" id="mw-searchButton" value="<?php $this->msg('searchbutton') ?>"<?php echo $this->skin->tooltipAndAccesskey( 'search-fulltext' ); ?> />
			</div>
		</form>

	</div> <!-- End header div -->

	<div id="mBody">

	    <!-- Dunno if this is important...? Who cares about IE5 anyways? C'mon, we got Firefox!

	    <script type="<?php $this->text('jsmimetype') ?>"> if (window.isMSIE55) fixalpha(); </script> -->

	    <!-- NAVIGATION -->
		<div id="side" class="noprint"> <!-- cavendish-mw / monobook: column-one -->
			<ul id="nav">
                <?php
                // Display Personal Tools block.
                $this->personalTools();

                // Display other Navigation blocks.
                $sidebar = $this->data['sidebar'];

                if ( !isset( $sidebar['SEARCH'] ) ) $sidebar['SEARCH'] = true;
                if ( !isset( $sidebar['TOOLBOX'] ) ) $sidebar['TOOLBOX'] = true;
                if ( !isset( $sidebar['LANGUAGES'] ) ) $sidebar['LANGUAGES'] = true;

                foreach ($sidebar as $boxName => $cont) {
                    if ( $boxName == 'SEARCH' ) {
                        // The searchbox is disabled, because we already have one in the header.
                        // Uncomment the line below to enable it again.
                        //$this->searchBox();
                    }
                    elseif ( $boxName == 'TOOLBOX' ) {
                        $this->toolbox();
                    }
                    elseif ( $boxName == 'LANGUAGES' ) {
                        $this->languageBox();
                    }
                    else {
                        $this->customBox( $boxName, $cont );
                    }
                }
                ?>
			</ul>
		</div><!-- end of SIDE div -->

        <!-- MAIN CONTENT -->
	    <div id="mainContent"> <!-- cavendish-mw / monobook: column-content -->
            <!-- <a name="top" id="top"></a> -->
            <!-- sitenotice was here -->
            <h1><?php $this->data['displaytitle']!=""?$this->html('title'):$this->text('title') ?></h1>
            <h3 id="siteSub"><?php $this->msg('tagline') ?></h3>
            <div id="contentSub"><?php $this->html('subtitle') ?></div>

            <?php if($this->data['undelete']) { ?><div id="contentSub2"><?php     $this->html('undelete') ?></div><?php } ?>
            <?php if($this->data['newtalk'] ) { ?><div class="usermessage"><?php $this->html('newtalk')  ?></div><?php } ?>
            <?php if($this->data['showjumplinks']) { ?><div id="jump-to-nav"><?php $this->msg('jumpto') ?> <a href="#column-one"><?php $this->msg('jumptonavigation') ?></a>, <a href="#searchInput"><?php $this->msg('jumptosearch') ?></a></div><?php } ?>
            <!-- start content -->
            <?php $this->html('bodycontent') ?>
            <?php if($this->data['catlinks']) { $this->html('catlinks'); } ?>
            <!-- end content -->
            <?php if($this->data['dataAfterContent']) { $this->html ('dataAfterContent'); } ?>
            <div class="visualClear"></div>
	    </div> <!-- End mainContent div -->

	</div> <!-- End mBody div -->

	<div class="visualClear"></div>

	<!-- FOOTER -->
    <div id="footer">
        <?php if($this->data['poweredbyico']) { ?>
            <div id="f-poweredbyico"><?php $this->html('poweredbyico') ?></div>

        <?php 	}
	    if($this->data['copyrightico']) { ?>
	        <div id="f-copyrightico"><?php $this->html('copyrightico') ?></div>

        <?php	} ?>

        <?php
	    // Generate additional footer links
	    $footerlinks = array(
		    //'lastmod', 'viewcount',
		    'numberofwatchingusers', 'credits', 'copyright',
		    'privacy', 'about', 'disclaimer', 'tagline',
	    );
	    ?>

	    <ul>
	    <?php
	    foreach( array('lastmod', 'viewcount') as $aLink ) {
		    if( isset( $this->data[$aLink] ) && $this->data[$aLink] ) { ?>
		        <li id="<?php echo $aLink ?>"><?php $this->html($aLink) ?></li>
        <?php } } ?>
        </ul>

        <?php
	    $validFooterLinks = array();
	    foreach( $footerlinks as $aLink ) {
		    if( isset( $this->data[$aLink] ) && $this->data[$aLink] ) {
			    $validFooterLinks[] = $aLink;
		    }
	    }

	    if ( count( $validFooterLinks ) > 0 ) { ?>
		    <ul id="f-list">
            <?php foreach( $validFooterLinks as $aLink ) {
			    if( isset( $this->data[$aLink] ) && $this->data[$aLink] ) { ?>
			    <li id="<?php echo $aLink ?>"><?php $this->html($aLink) ?></li>
            <?php } } ?>
            <li class="noprint"><a href="https://sourceforge.net/projects/cavendishmw/">Cavendish Skin</a></li>
		    </ul>
        <?php	} ?>
    </div> <!-- End of footer div -->

</div> <!-- End of container div -->

<?php $this->html('bottomscripts'); /* JS call to runBodyOnloadHook */ ?>
<?php $this->html('reporttime') ?>
<?php if ( $this->data['debug'] ): ?>
<!-- Debug output:
<?php $this->text( 'debug' ); ?>

-->
<?php endif; ?>
</body></html>
<?php
wfRestoreWarnings();
} // end of execute() method

/*************************************************************************************************/
function personalTools() {
?>
				<li><span><?php $this->msg('personaltools') ?></span>
				    <ul>
			            <?php foreach($this->data['personal_urls'] as $key => $item) { ?>
                        <li id="pt-<?php echo htmlspecialchars($key) ?>"><a href="<?php echo htmlspecialchars($item['href']) ?>" <?php if(!empty($item['class'])) { ?>
                            class="<?php echo htmlspecialchars($item['class']) ?>"<?php } ?>><?php echo htmlspecialchars($item['text']) ?></a></li>
	                    <?php } ?>
				    </ul>
				</li>
<?php
}
function searchBox() {
	global $wgUseTwoButtonsSearchForm;
?>
<li><span><label for="searchInput"><?php $this->msg('search') ?></label></span>
    <ul>
        <form action="<?php $this->text('wgScript') ?>" id="searchform">
        <input type='hidden' name="title" value="<?php $this->text('searchtitle') ?>"/>
        <input id="searchInput" name="search" type="text"<?php echo $this->skin->tooltipAndAccesskey('search');
        if( isset( $this->data['search'] ) ) { ?> value="<?php $this->text('search') ?>"<?php } ?> />
        <div>
            <input type='submit' name="go" class="searchButton" id="searchGoButton"	value="<?php $this->msg('searcharticle') ?>"<?php echo $this->skin->tooltipAndAccesskey( 'search-go' ); ?> />
            <?php if ($wgUseTwoButtonsSearchForm) { ?>
                <input type='submit' name="fulltext" class="searchButton" id="mw-searchButton" value="<?php $this->msg('searchbutton') ?>"<?php echo $this->skin->tooltipAndAccesskey( 'search-fulltext' ); ?> />
            <?php } else { ?>
                <a href="<?php $this->text('searchaction') ?>" rel="search"><?php $this->msg('powersearch-legend') ?></a>
            <?php } ?>
        </div>
        </form>
    </ul>
</li>
<?php
	}

	/*************************************************************************************************/
	function toolbox() {
?>
<li><span><?php $this->msg('toolbox') ?></span>
    <ul>
        <?php if($this->data['notspecialpage']) { ?>
            <li id="t-whatlinkshere"><a href="<?php echo htmlspecialchars($this->data['nav_urls']['whatlinkshere']['href']) ?>"<?php echo $this->skin->tooltipAndAccesskey('t-whatlinkshere') ?>><?php $this->msg('whatlinkshere') ?></a></li>

        <?php if( $this->data['nav_urls']['recentchangeslinked'] ) { ?>
		    <li id="t-recentchangeslinked"><a href="<?php echo htmlspecialchars($this->data['nav_urls']['recentchangeslinked']['href']) ?>"<?php echo $this->skin->tooltipAndAccesskey('t-recentchangeslinked') ?>><?php $this->msg('recentchangeslinked') ?></a></li>
        <?php
        } }

		if(isset($this->data['nav_urls']['trackbacklink'])) { ?>
		    <li id="t-trackbacklink"><a href="<?php echo htmlspecialchars($this->data['nav_urls']['trackbacklink']['href']) ?>"<?php echo $this->skin->tooltipAndAccesskey('t-trackbacklink') ?>><?php $this->msg('trackbacklink') ?></a></li>
        <?php
        }

		if($this->data['feeds']) { ?>
		    <li id="feedlinks"><?php foreach($this->data['feeds'] as $key => $feed) { ?><a id="<?php echo Sanitizer::escapeId( "feed-$key" ) ?>" href="<?php echo htmlspecialchars($feed['href']) ?>" rel="alternate" type="application/<?php echo $key ?>+xml" class="feedlink"<?php echo $this->skin->tooltipAndAccesskey('feed-'.$key) ?>><?php echo htmlspecialchars($feed['text'])?></a>&nbsp;<?php } ?></li>
		<?php
		}

		foreach( array('contributions', 'log', 'blockip', 'emailuser', 'upload', 'specialpages') as $special ) {
			if($this->data['nav_urls'][$special]) { ?>
				<li id="t-<?php echo $special ?>"><a href="<?php echo htmlspecialchars($this->data['nav_urls'][$special]['href']) ?>"<?php echo $this->skin->tooltipAndAccesskey('t-'.$special) ?>><?php $this->msg($special) ?></a></li>
        <?php
        } }

		if(!empty($this->data['nav_urls']['print']['href'])) { ?>
            <li id="t-print"><a href="<?php echo htmlspecialchars($this->data['nav_urls']['print']['href']) ?>" rel="alternate"<?php echo $this->skin->tooltipAndAccesskey('t-print') ?>><?php $this->msg('printableversion') ?></a></li><?php
		}

		if(!empty($this->data['nav_urls']['permalink']['href'])) { ?>
            <li id="t-permalink"><a href="<?php echo htmlspecialchars($this->data['nav_urls']['permalink']['href'])?>"<?php echo $this->skin->tooltipAndAccesskey('t-permalink') ?>><?php $this->msg('permalink') ?></a></li>
        <?php } elseif ($this->data['nav_urls']['permalink']['href'] === '') { ?>
			<li id="t-ispermalink"<?php echo $this->skin->tooltip('t-ispermalink') ?>><?php $this->msg('permalink') ?></li>
	    <?php
	    }

	    /* This line seems unimportant for this skin
	    wfRunHooks( 'MonoBookTemplateToolboxEnd', array( &$this ) );
	    */

	    // This line triggers that certain extensions can add extra links to the toolbox.
	    wfRunHooks( 'SkinTemplateToolboxEnd', array( &$this ) );
        ?>
	</ul>
</li>
<?php
	}

	/*************************************************************************************************/
	function languageBox() {
		if( $this->data['language_urls'] ) {
?>
<li><span><?php $this->msg('otherlanguages') ?></span>
    <ul>
        <?php foreach($this->data['language_urls'] as $langlink) { ?>
		<li class="<?php echo htmlspecialchars($langlink['class'])?>"><a href="<?php echo htmlspecialchars($langlink['href']) ?>"><?php echo $langlink['text'] ?></a></li>
        <?php } ?>
	</ul>
</li>
<?php
		}
	}

	/*************************************************************************************************/
	function customBox( $bar, $cont ) {
?>
<li><span><?php $out = wfMsg( $bar ); if (wfEmptyMsg($bar, $out)) echo $bar; else echo $out; ?></span>
    <?php if ( is_array( $cont ) ) { ?>
    <ul>
        <?php foreach($cont as $key => $val) { ?>
	    <li id="<?php echo Sanitizer::escapeId($val['id']) ?>"<?php if ( $val['active'] ) { ?> class="active" <?php } ?>><a href="<?php echo htmlspecialchars($val['href']) ?>"<?php echo $this->skin->tooltipAndAccesskey($val['id']) ?>><?php echo htmlspecialchars($val['text']) ?></a></li>
        <?php } ?>
    </ul>
    <?php } else {
		# allow raw HTML block to be defined by extensions
		print $cont;
	}
?>
</li>
<?php
}

} // end of class


